mne_bids/utils.py

```diff
@@ -371,31 +375,62 @@ def _find_matching_sidecar(bids_fname, bids_root, suffix, allow_fail=False):
         and no sidecar_fname was found
 
     """
-    bids_fname = op.basename(bids_fname)
+    params = _parse_bids_filename(bids_fname, verbose=False)
 
     # We only use subject and session as identifier, because all other
     # parameters are potentially not binding for metadata sidecar files
-    if 'ses' in bids_fname:
-        search_str = '_'.join(bids_fname.split('_')[:2])
-    else:
-        # only sub, no ses
-        search_str = '_'.join(bids_fname.split('_')[:1])
+    search_str = 'sub-' + params['sub']
+    if params['ses'] is not None:
+        search_str += '_ses-' + params['ses']
 
-    # find the sidecar file, doing a recursive glob from the bids_root
+    # Find all potential sidecar files, doing a recursive glob from bids_root
     search_str = op.join(bids_root, '**', search_str + '*' + suffix)
     candidate_list = glob.glob(search_str, recursive=True)
-    if len(candidate_list) != 1:
+
+    # Assign each candidate a score, based on how many entities are shared with
+    # bids_fname. The candidate with the highest score wins. Candidates with
+    # entities that conflict with bids_fname are disqualified.
+    best_candidates = []
+    best_score = 0
+    for candidate in candidate_list:
+        n_matches = 0
+        candidate_disqualified = False
+        candidate_params = _parse_bids_filename(candidate, verbose=False)
+        for entity, value in params.items():
+            if value is None:
+                continue
+            if entity in candidate_params:
+                if candidate_params[entity] is None:
+                    continue
+                elif candidate_params[entity] == value:
+                    n_matches += 1
+                else:
+                    # Incompatible entity found, candidate is disqualified
+                    candidate_disqualified = True
+                    break
+        if not candidate_disqualified and n_matches >= best_score:
+            best_score = n_matches
+            best_candidates.append(candidate)
```

Comment:
If on the first run through the loop, I find a file with `n_matches=3` ... and on the second run through the loop, I find one with `n_matches=4`

... then I have two `best_candidates` ... but they did not tie (the second was better) --> I think the code is currently not covering that, is it?

Reply:
doh!